### ISSUE TITLE
[Feature Request] Patch Picker

### ISSUE THREAD

**Issue by brianblakely** [libretro/RetroArch#16378]
## Background

Softpatching represents a significant advancement in RetroArch, shifting the burden of managing content patches away from the file system. However, patch releases have become so popular that you might have *dozens* of mutually exclusive patches for a single piece of content. This becomes cumbersome to manage, as you must rename or move files around to choose which patch is active.

### Requested Feature

Enable users to place patches in the file system one time and then, within RetroArch, select which patches to apply to content before running it.

### Proposed Usage

1. Adopt an extension format `content.patch-name.ips[N]`, where `.patch-name.` is a user-defined key, distinguishing each patch or group of patches. Ex: `My Content.Patch Name.ips, My Content.Patch Name.ips1, My Content.Another Patch.ips`
2. Within the content view—the interface for setting core associations, adding favorites, etc.—introduce a "Set Patch" menu item. This feature would enable users to choose a patch/patch group from a list based on the `patch-name` key. Perhaps in the future, a patch hash database could pretty-print patch names automatically.
3. The menu item would also allow a "No Patch" option, making it possible to run the unpatched content, which would be valuable for both users and patch developers in testing.

**Comment by hizzlekizzle** [comment]
When we've discussed this in the past, the preferred workflow was: name a directory after the ROM and fill it with patches, then add a 'patch' (or 'patch-and-run') entry to the existing 'run' menu where you can choose which to apply.

So, if you have Foo.sfc and Bar.sfc in a directory, you would have a structure like this:

```
Roms
-|Foo.sfc
-|Foo
--|patch1.ips
--|patch2.ips
--|etc.
-|Bar.sfc
-|Bar
--|patch1.bps
--|patch2.bps
--|etc.
```

**Comment by brianblakely** [comment]
That's terrific, and I think it makes sense for saves/states to integrate with this system as well, as patches can often lead to confusing or harmful results for save management.

**Comment by i30817** [comment]
As long as multiple patches automatic application doesn't get broken (very common in romhacking.net addendum patches)...

**Comment by brianblakely** [comment]
@i30817 The existing "workflow" shouldn't change at all, and directory workflow allows for it as well. Example of multiple patch groups living alongside one another, as well as alongside individual patches:

```
Roms
-|Foo.sfc
-|Foo
--|My Mod.ips
--|My Mod.ips1
--|My Mod.ips2
--|Another Mod.ips
--|Another Mod.ips1
--|Yet Another Mod.ips
--|And So On.ips
```

**Comment by hizzlekizzle** [comment]
Yeah, whether and/or how a hypothetical patch-picker and the existing multi-patch system interact would be up to the implementer to sort out. Sometimes that stuff gets complicated and it's better to just punt and not have them interact at all at first. No way to know until someone starts working on it.

**Comment by brianblakely** [comment]
@hizzlekizzle I am going to take a look into it over the weekend and see if I can pick this up, unless someone else hits it first.

**Comment by i30817** [comment]
Don't forget that unfortunately some rom formats cue\bin for instance, already have multiple formats with the same name minus extension, so the patch directory might have to not be exactly the same name as the rom minus extension because of target confusion. This is currently not relevant for cue\bin but might be in the future and there are always other formats using the idea so likely something like this is already happening such as audio extensions for sfc roms or bnes xml formats or something...

The same way, it can't be exactly the same name because filesystems won't allow it.

tl;Dr: you might need a suffix in the directory name.

**Comment by hizzlekizzle** [comment]
yeah, fair points. [gamename]-patches or somesuch might be better.

**Comment by i30817** [comment]
I really hated typing that comment because something like 'Super Metroid (USA).sfc-patches' is aesthetically displeasing to say the least, but RetroArch is so wide because of the number of emulator cores under its umbrella that its the only thing I could think of that wouldn't conflict or cause some target confusion. 

Maybe someone else has a better idea.

**Comment by i30817** [comment]
Are you saying you plan to add the field to the playlist JSON file entries?

Thats going to interfere with the rescan playlist functionality.

.

**Comment by brianblakely** [comment]
@i30817 
> Thats seems like its going to interfere with the rescan playlist functionality. You might have to fix bugs there to prevent a rescan\recreation of a playlist not to nuke a saved patch.

Would you kindly provide some steps to test this?

**Comment by i30817** [comment]
Let's see. The way the manual scanner does this is a bit weird.

A playlist created by the manual scanner has various extra fields compared to a normal scan one.

The relevant ones being scan_content_dir (activates the ability to refresh a playlist in the playlist management menu or if you press start\long android press while focused in a playlist name in ozone iirc) and the scan_* fields. Which are properties/options of the scan. One relevant one is scan_overwrite_playlist or not. 

The difference is if it is false, the refresh option just adds new accepted rom entries, and if it is true, the playlist is first deleted.

You usually WANT to overwrite the playlist to not have to click the other option 'clean playlist' which deletes nonexistent entries from a playlist. Anyway, the overwrite playlist option will probably nuke those extra properties.

**Comment by i30817** [comment]
There is also the fact that for some inexplicable and probably dumb reason RetroArch still supports the old style playlists that don't support custom properties (it's a awful line format with a truly atrocious way to tell when the entries stop and playlist properties start... It involves % 6 enough said).

And not just reading, there is a actual option to write it too.

**Comment by brianblakely** [comment]
@i30817 Thank you very much for that description, I fully understand now. It may be out of scope for this issue to resolve UX gaps in the Manual Scan feature. The idea is to follow the same UX pattern as custom labels (i.e. "Rename" content option). Custom labels are indeed wiped out when a Manual Scan playlist is overwritten. However, if/when Manual Scan behavior changes, it should be possible for patches to be preserved just as well as custom labels.

**Comment by i30817** [comment]
Changing this would affect that, and fixing that would be a opportunity to fix  other things. Although fixing the worst problem of playlists (renaming one sabotaging thumbnails, manual playlist or not) would be a radical departure, that would finally kill off the older playlist format for good, probably.

**Comment by i30817** [comment]
Btw your solution wouldn't work for a hypothetical future where RetroArch allows cd softpatching (because tracks are not part of the playlist, only 'index files' (cue, toc, m3u, iso would work because it's a single file format, although isos are rare).

Fortunately, if the emulator supports chd, and do the bare minimum work to support parent-child discovery, there is already a alternative to softpatch cds...

**Comment by i30817** [comment]
Mmm? The problem is that when RetroArch loads a file to send to a core, then tries to patch it before, it would only get a cue file (sometimes a m3u too for one more level of indirection). It's the cores that then load tracks, as far as I know.

**Comment by Zal0** [comment]
I recently submited a PR based on a few things mentioned here

Will be happy to see your thoughts about it

https://github.com/libretro/RetroArch/pull/18825
